cmd/compose/port_cleanup.go

```diff
@@ -0,0 +1,44 @@
+package compose
+
+import (
+	"fmt"
+	"os/exec"
+	"strings"
+)
+
+func killProcessesUsingExposedPorts(project *types.Project) error {
+	for _, service := range project.Services {
+		for _, port := range service.Ports {
+			exposedPort := port.Published
+			if exposedPort != "" {
+				err := killProcessOnPort(exposedPort)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func killProcessOnPort(port string) error {
+	cmd := exec.Command("lsof", "-t", "-i", fmt.Sprintf(":%s", port))
+	output, err := cmd.Output()
+	if err != nil {
+		return fmt.Errorf("failed to find process using port %s: %v", port, err)
+	}
+
+	pid := strings.TrimSpace(string(output))
+	if pid == "" {
+		return nil
+	}
+
+	killCmd := exec.Command("kill", "-9", pid)
```

Comment:
This part of the code is not portable and won't work on Windows environment for example.
The fix proposed should work for Linux, Mac and Windows

Reply:
Also Compose is a client library, while docker engine may run on a remote machine, so this isn't the relevant place to propose such a fix. We already invoke `ContainerStop` API which is responsible to stop (or kill after timeout) the target container and release resources, if you found a bug with this, better report on github.com/moby/moby